app/models/package_manager/base/bulk_version_updater.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module PackageManager
+  class Base
+    # This class is responsible for taking a project and a list of ApiVersions (aka raw versions mapped
+    # to a common set of attributes), and creating Versions for the Project in the database, in a single
+    # query. Because we're using +upsert_all()+, Versions that already exist will just be updated.
+    # This single query avoids all the N+1s involved in a regular batch insert of Versions that we did in the past.
+    class BulkVersionUpdater
+      def initialize(db_project:, api_versions:, repository_source_name:)
+        @db_project = db_project
+        @api_versions = api_versions
+        @repository_source_name = repository_source_name
+      end
+
+      def run!
+        return if @api_versions.empty?
+
+        # create synthetic versions without saving them, so we can get their attributes
+        attrs = @api_versions
+          .map { |api_version| Version.new(api_version.to_version_model_attributes.merge(project: @db_project)) }
```

Comment:
Thoughts about doing a bulk find_or_create on the versions here instead? One concern is that while the synthetic objects might be valid when constructed here from scratch, but they might not be valid when you do the upsert due to existing state. 

For example, if an existing row has an invalid status, I think the from-scratch version would pass validation here (no changes to status), but after the upsert it would still have an invalid status, so you'd end up with a Version that doesn't pass validation (if the intention here is to only save valid Version objects).

Reply:
the intention here is just to make sure we run validations and callbacks on each Version beforehand, as if they were inserted individually. There's some logic on line 43 to prefer the new data if possible (`status = EXCLUDED.status,`) so that'll squash the invalid status case.